api/src/schemas.py

```diff
@@ -230,49 +230,58 @@ class Config:
+            )
+        if not values["is_periodic"] and values["time_density"] != None:
+            raise ValueError(
+                "Invalid time density - if is_periodic is false, time_density must be null"
+            )
+        return values
 
     # collection id must be all lowercase, with optional - delimiter
     @validator("collection")
-    def check_id(cls, v):
-        if not re.match(r"[a-z]+(?:-[a-z]+)*", v):
-            raise ValueError("Invalid id")
-        return v
+    def check_id(cls, collection):
+        if not re.match(r"[a-z]+(?:-[a-z]+)*", collection):
+            raise ValueError(
+                "Invalid id - id must be all lowercase, with optional '-' delimiters"
+            )
+        return collection
 
     # all sample files must begin with prefix and their last element must match regex
     @root_validator
-    def check_sample_files(cls, v):
-        if "s3" not in [item.discovery for item in v["discovery_items"]]:
+    def check_sample_files(cls, values):
+        if "s3" not in [item.discovery for item in values["discovery_items"]]:
             print("No s3 discovery items to validate sample files against")
-            return v
+            return values
         # TODO cmr handling/validation
+
         valid_matches = []
-        for item in v["discovery_items"]:
+        for item in values["discovery_items"]:
             if item.discovery == "s3":
                 valid_matches.append(
                     {"prefix": item.prefix, "regex": item.filename_regex}
                 )
-        for fname in v["sample_files"]:
-            if not any([fname.startswith(match["prefix"]) for match in valid_matches]):
-                raise ValidationError(
-                    f"Invalid sample file - {fname} doesn't match prefix"
-                )
-            if not any(
-                [
-                    re.search(match["regex"], fname.split("/")[-1])
-                    for match in valid_matches
-                ]
-            ):
-                raise ValidationError(
-                    f"Invalid sample file - {fname} doesn't match provided regex"
-                )
-        return v
+
+        invalid_fnames = []
+        for fname in values["sample_files"]:
+            prefix_matches = [
+                fname.startswith(match["prefix"]) for match in valid_matches
+            ]
+            regex_matches = [
+                re.search(match["regex"], fname.split("/")[-1])
+                for match in valid_matches
+            ]
+            if not any([a and b for a, b in zip(prefix_matches, regex_matches)]):
+                invalid_fnames.append(fname)
+        if invalid_fnames:
+            raise ValidationError(
+                f"Invalid sample file(s) - {invalid_fnames} do not match any of the provided prefix/filename_regex combinations"
+            )
+        return values
```

Comment:
this seems too complicated 😅 
something like this: ?
```python
valid_matches = {}

for item in values["discovery_items"]:
    if item["discovery"] == "s3":
        valid_matches.setdefault(item["prefix"].strip("/"), []).append(item["filename_regex"])

invalid_fnames = []

for fname in values["sample_files"]:
    prefix, name = fname.rsplit("/", 1)
    if file_regex := valid_matches.get(prefix):
        matches = [re.search(reg, name) for reg in file_regex]
        if not any(matches):
            invalid_fnames.append(fname)
    else:
        invalid_fnames.append(fname)

print(f"{invalid_fnames=}")
```

Reply:
I like the refactor, but I think those changes are from an older commit. I added part of your proposed changes to the PR though.